### PR TITLE
[Addons] Ensure addon installation order when migrating to matrix

### PR DESCRIPTION
**Pull request by enen92**
## Description
Despite our efforts to ensure that every single addon in the matrix branch has a higher version than those stored in lower branches (so they all get updated when migrating from leia to matrix), the reality is that right now almost all of them get disabled and won't update.

This happens because kodi doesn't enforce any control in the order of addon installations. So let's say that you have `plugin.video.foo` and `plugin.video.bar`, and `foo` depends on `bar`, if `foo` gets updated and installed before `bar` it can lead to `foo` getting disabled. This used to work just fine before matrix because if `foo` was updated before `bar`, as the installed version of `bar` was still compatible with kodi (`xbmc.python` was backwards compatible) `foo` was updated and installed and later `bar` was also updated and installed. In matrix, even though `foo` requires a minimum version of `bar` equal to the one installed before starting the migration, `bar` is invalid because it depends on a `xbmc.python` version lower than the minimum supported by kodi matrix - leading to almost any addon with dependencies to get disabled.

So, this PR does the following:
- Ensure that concurrency is avoided in bulk addon updates: if the repo manifest is updated while migrating, it will triggered another round of updates while the first one is still running
- Ensure the `InstallUpdates` function can be configured to wait for each addon to be installed before requesting the installation of another one (using only in the migration phase - first time you run a new kodi version)
- Sort the list of addons to be updated depending on their dependencies. Addons that depend on others should be the last ones to be installed.

Note that this only affects the addon migration, i.e, when running matrix for the first time. All the other updates (after first run) will be exactly the same, the list is sorted but they are installed without any order - this is faster and won't cause any issues since the version being updated is already "new-kodi" compatible.

The only drawback of this approach is that the addon migration is now a bit slower, but I guess we can live with that.

@AlwinEsch for the review
@da-anda @ronie @anxdpanic fyi

Thanks to @yol for the c++ tips

## Motivation and Context
Promote a seamless migration from Leia to Matrix for users despite all the "non-visible" work and fussle introduced with the python2 -> python3 migration

## How Has This Been Tested?
- Install kodi leia
- Install addons in leia that are known to also be in matrix
- close kodi leia
- run matrix
- check none is disabled (and if any is, it's our fault managing the repo)

## Screenshots (if appropriate):
**Leia:**
Leia

**Master/Matrix:**
Matrix

**PR:**
Leia

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/AddonInstaller.h [hunk 1]
```diff
@@ -135,6 +146,8 @@ class CAddonInstaller : public IJobCallback
   mutable CCriticalSection m_critSection;
   JobMap m_downloadJobs;
   CEvent m_idle;
+  // controls whether Kodi is already updating addons
+  std::atomic<bool> m_updatingAddons{false};
```
Comment (AlwinEsch) [inline xbmc/addons/AddonInstaller.h:150]: Seen my runtime test, that the atomic makes this and need above the `AddonInstaller.h` a `#include <atomic>` by me on Linux.:
```
In file included from /home/alwin/Development/Kodi/kodi-Matrix/xbmc/games/dialogs/GUIDialogSelectGameClient.cpp:12:
/home/alwin/Development/Kodi/kodi-Matrix/xbmc/addons/AddonInstaller.h:150:21: error: field ‘m_updatingAddons’ has incomplete type ‘std::atomic<bool>’
  150 |   std::atomic<bool> m_updatingAddons{false};
      |                     ^~~~~~~~~~~~~~~~
In file included from /usr/include/c++/9/bits/shared_ptr_atomic.h:33,
                 from /usr/include/c++/9/memory:82,
                 from /home/alwin/Development/Kodi/kodi-Matrix/xbmc/games/GameTypes.h:11,
                 from /home/alwin/Development/Kodi/kodi-Matrix/xbmc/games/dialogs/GUIDialogSelectGameClient.h:11,
                 from /home/alwin/Development/Kodi/kodi-Matrix/xbmc/games/dialogs/GUIDialogSelectGameClient.cpp:9:
/usr/include/c++/9/bits/atomic_base.h:145:12: note: declaration of ‘struct std::atomic<bool>’
  145 |     struct atomic;
      |            ^~~~~~
```

**Review by AlwinEsch — APPROVED** [review]
From code and work it looks good also was the test OK.
Think there a lot of another parts where need fixed (on other requests), but this is a good first step.
Only a bit of docs and one compile fail fix for Linux needed.

:+1:

**Comment by Rechi** [comment]
jenkins build this please

**Comment by yol** [comment]
I actually think it would be better to always do this. Suppose you update an addon and it requires a newer version of a dependency. Let's say it will fail to load if the dependency is too old (e.g. because it imports a symbol that is not available in the old version) - wouldn't you hit the same problem if the addon gets installed before the dependency has finished installing?

File: xbmc/addons/AddonInstaller.cpp [hunk 2]
```diff
@@ -413,24 +413,83 @@ void CAddonInstaller::PrunePackageCache()
...
+      const auto& addon = *it;
+
+      auto dependencies = addon->GetDependencies();
+      bool addToSortedList = true;
+      // if the addon has dependencies we need to check for each dependency if it also has
+      // an update to be installed (and in that case, if it is already in the sorted vector).
+      // if all dependency match the said conditions, the addon doesn't depend on other addons
+      // waiting to be updated. Hence, the addon being processed can be installed (i.e. added to
+      // the end of the sorted vector of addon updates)
+      if (!dependencies.empty())
```
Comment (yol) [inline xbmc/addons/AddonInstaller.cpp:458]: you can leave out the whole if block

File: xbmc/addons/AddonInstaller.cpp [hunk 3]
```diff
@@ -413,24 +413,83 @@ void CAddonInstaller::PrunePackageCache()
...
+      {
+        for (const auto& dep : dependencies)
+        {
+          auto comparator = [dep](const std::shared_ptr<ADDON::IAddon> addon) {
```
Comment (yol) [inline xbmc/addons/AddonInstaller.cpp:462]:
```suggestion
          auto comparator = [&dep](const std::shared_ptr<ADDON::IAddon>& addon) {
```

File: xbmc/addons/AddonInstaller.cpp [hunk 4]
```diff
@@ -413,24 +413,83 @@ void CAddonInstaller::PrunePackageCache()
...
+void CAddonInstaller::SortByDependencies(std::vector<std::shared_ptr<ADDON::IAddon>>& updates)
+{
+  std::vector<std::shared_ptr<ADDON::IAddon>> sorted;
+  while (!updates.empty())
+  {
+    for (auto it = updates.begin(); it != updates.end();)
+    {
+      const auto& addon = *it;
+
+      auto dependencies = addon->GetDependencies();
```
Comment (yol) [inline xbmc/addons/AddonInstaller.cpp:451]:
```suggestion
      const auto& dependencies = addon->GetDependencies();
```

**Review by yol — CHANGES_REQUESTED** [review]
some minors

**Comment by enen92** [comment]
@yol making this default will certainly avoid issues like that one. Reason I didn't make it default is to try to change as less as possible so it won't cause future issues we can't see right now (seems a classic when doing kodi stuff :) ).

Either way, this needs further work. We shouldn't be guarding the `InstallUpdates` but the migration process instead.

**Comment by yol** [comment]
@enen92 In general, I agree. I also try to keep changes as local as possible. But not resolving addon installation order correctly (or at all) is a serious flaw that will bite us again. So IMO the benefits outweigh here.

**Comment by phunkyfish** [comment]
Why not ask @MilhouseVH to put it in a build for a few weeks.

If no issues are raised then it can becomes the standard route for any addon update cycle.

**Comment by enen92** [comment]
Made a few changes to the PR:
- We're now guarding the migration process so that the updates doesn't trigger while the migration is happening (and not when updates are hapening like before)
- Made the process of updates ordered by default as per @yol suggestion

Not related to the issue this pr tries to solve:
@AlwinEsch do you think it also makes sense to block addon updates when another process is also updating addons? Kodi triggers addon updates whenever a repo manifest is updated. So (unless in the migration phase were we specifically wait for all) if you have more than 1 repo installed it is not guaranteed that all addons in that repo will be able to update if they contain (incompatible) dependencies to addons in other repos.  To be honest, I'm not even sure if this is a problem we should solve. There have been discussions in the past regarding external repos/addons leeching our infrastructure. Maybe something to be considered in https://github.com/xbmc/xbmc/pull/17677 @matthuisman ?

File: xbmc/addons/AddonSystemSettings.cpp [hunk 5]
```diff
@@ -118,8 +118,12 @@ bool CAddonSystemSettings::UnsetActive(const AddonPtr& addon)
   return true;
 }
 
+bool CAddonSystemSettings::isMigratingAddons() const
```
Comment (yol) [inline xbmc/addons/AddonSystemSettings.cpp:121]:
```suggestion
bool CAddonSystemSettings::IsMigratingAddons() const
```

always use PascalCase

File: xbmc/addons/AddonSystemSettings.h [hunk 6]
```diff
@@ -41,10 +42,14 @@ class CAddonSystemSettings : public ISettingCallback
...
-   * @param onMigrate Called when a long running migration task takes place.
    * @return list of addons that was modified.
    */
-  std::vector<std::string> MigrateAddons(std::function<void(void)> onMigrate);
+  std::vector<std::string> MigrateAddons();
+
+  /*!
+   * Checks if Kodi is migrating addons
+   */
+  bool isMigratingAddons() const;
```
Comment (yol) [inline xbmc/addons/AddonSystemSettings.h:52]:
```suggestion
  bool IsMigratingAddons() const;
```

**Review by yol — CHANGES_REQUESTED** [review]
minors again :) behavior-wise I have no complaints :+1:

**Comment by MilhouseVH** [comment]
> Why not ask @MilhouseVH to put it in a build for a few weeks.

Just being extra cautious, but I'd like to see this bake for another 24 hours and then I'd be happy to include it in `#0421` assuming nobody finds anything of concern etc.

**Comment by enen92** [comment]
@MilhouseVH please do not add it yet (and probably also not in 24 hours), I still want to work a bit more on this (note the wip label), and likely will require some refactor pr before continuing this one.

**Comment by MilhouseVH** [comment]
OK no problem, ping me when you think it's ready!

File: xbmc/addons/Repository.h [hunk 7]
```diff
@@ -70,10 +70,13 @@ namespace ADDON
     explicit CRepositoryUpdateJob(const RepositoryPtr& repo);
     ~CRepositoryUpdateJob() override = default;
     bool DoWork() override;
+    void ShouldInstallAddons(const bool install) { m_shouldinstalladdons = install; };
```
Comment (phunkyfish) [inline xbmc/addons/Repository.h:73]: Maybe rename to `SetShouldInstallAddons`? Also, for basic types the const is not really required (but it does no harm).

File: xbmc/addons/RepositoryUpdater.cpp [hunk 8]
```diff
@@ -170,7 +174,7 @@ void CRepositoryUpdater::OnTimeout()
   }
 
   CLog::Log(LOGDEBUG,"CRepositoryUpdater: running scheduled update");
-  CheckForUpdates();
+  CheckForUpdates(false, true);
```
Comment (phunkyfish) [inline xbmc/addons/RepositoryUpdater.cpp:177]: The second argument forces install to true. Should it use the `install` variable instead?

**Review by phunkyfish — CHANGES_REQUESTED** [review]

File: xbmc/addons/AddonManager.h [hunk 9]
```diff
@@ -281,6 +283,33 @@ namespace ADDON
...
+     * \param wait if the process should wait for all addons to install
+     */
+    void InstallAddonUpdates(VECADDONS& updates, bool wait) const;
+
+    // This guards the addon installation process to make sure
+    // addon updates are not installed concurrently
+    // while the migration is running. Addon updates can be triggered
+    // as a result of a repository update event.
+    // (migration will install any available update anyway)
+    mutable std::mutex m_installaddonsmutex;
```
Comment (phunkyfish) [inline xbmc/addons/AddonManager.h:311]: `m_installAddonsMutex`

File: xbmc/addons/AddonManager.cpp [hunk 10]
```diff
@@ -410,14 +415,74 @@ std::vector<std::string> CAddonMgr::MigrateAddons()
...
+}
+
+bool CAddonMgr::GetAddonUpdateCandidates(VECADDONS& updates) const
 {
   // Get Addons in need of an update and remove all the blacklisted ones
-  auto updates = GetAvailableUpdates();
-  updates.erase(std::remove_if(updates.begin(), updates.end(), [this](const AddonPtr& addon) {
-    return IsBlacklisted(addon->ID());
-  }), updates.end());
+  auto updatesAll = GetAvailableUpdates();
```
Comment (phunkyfish) [inline xbmc/addons/AddonManager.cpp:428]: Why not just use `updates` here?

**Review by phunkyfish — CHANGES_REQUESTED** [review]

**Comment by enen92** [comment]
@MilhouseVH I think it is ready now. Can you include this on the next nightly?

**Comment by AlwinEsch** [comment]
> @AlwinEsch do you think it also makes sense to block addon updates when another process is also updating addons? 
For now yes, if the blocking not result in a freeze of something. 
In future there maybe it would be better to have a more intelligent processing about, thought about a centralized part/thread where do it and where the others register his works, this then check how is used and inform related places about. And then in makes all changes together at same time.

**Comment by MilhouseVH** [comment]
> @MilhouseVH I think it is ready now. Can you include this in the next nightly?

Well include in `#0426`.

**Review by yol — APPROVED** [review]

**Review by phunkyfish — APPROVED** [review]
Well done, your first threading c++ badge ;)

File: xbmc/addons/AddonManager.h [hunk 11]
```diff
@@ -280,6 +280,12 @@ namespace ADDON
     bool EnableSingle(const std::string& id);
 
     void FindAddons(ADDON_INFO_LIST& addonmap, const std::string& path);
+
```
Comment (Rechi) [inline xbmc/addons/AddonManager.h]: remove the white space at the end of the line

**Comment by enen92** [comment]
@MilhouseVH any complains so far (apart from possible ones caused by halifax downtime)?

**Pull request merged**